### PR TITLE
Initial pandas.typing Module

### PR DESCRIPTION
**Pull request by WillAyd**
ref #25791 maybe more of a conversation starter than complete PR but wanted to stub out some types that appear commonly in pandas code-base.

Haven't used these yet as wanted to agree on type first. I wouldn't be surprised if we start seeing some issues pop up that would require real development changes once these get utilized

@jreback @jbrockmendel

### CODE REVIEW

**Comment by jreback** [comment]
cool, can you use these in a couple of places? and some examples / docs-strings?

File: pandas/typing.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+from datetime import datetime, timedelta
+from pathlib import Path
+from typing import AnyStr, IO, Union
+
+from pandas._libs.tslibs import Period, NaT, Timedelta, Timestamp
+
+
+DateTimeLike = Union[datetime, timedelta, Period, Timedelta, Timestamp]
```
Comment (jbrockmendel) [inline pandas/typing.py]: Do/should the numpy versions go in here?

**Comment by WillAyd** [comment]
I don’t think so but not sure I follow - what’s the link between NumPy version and types?

**Comment by jbrockmendel** [comment]
> I don’t think so but not sure I follow - what’s the link between NumPy version and types?

Bad phrasing on my part.  I was referring to np.datetime64 and np.timedelta64 objects.  Many of our functions that accept datetime inputs will also accept np.datetime64 inputs.  Ditto for timedelta.

**Comment by WillAyd** [comment]
Sure just added the FilePathOrBuffer type to io.parsers. CI will fail with the following:

```sh
pandas/io/parsers.py:445: error: Item "str" of "Union[str, Path, IO[Any]]" has no attribute "close"
pandas/io/parsers.py:445: error: Item "Path" of "Union[str, Path, IO[Any]]" has no attribute "close"
```

There's an implicit assumption in the codebase that a local variable `should_close` dictates whether or not a `close` method gets called, which Mypy doesn't like. I suppose that is rather fragile so refactor could be inline

@jbrockmendel any thoughts on where best to try out the DateTimeLike types?

**Comment by jbrockmendel** [comment]
> any thoughts on where best to try out the DateTimeLike types?

Not really.  I guess I'm not clear on the use case.  Aside from running mypy, are we actually doing anything with these?

**Comment by WillAyd** [comment]
Use case would be to improve readability and maintainability of code base and in theory reducing bugs. There's some refactor that will go hand in hand with it (as you might have seen with some of the IO stuff above).

This is just starting things off and there's a long way to go. I think I might split off the datetimelike stuff for this for now and just focus on IO. Can see how that goes and learn from there

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -409,7 +410,12 @@ def _read(filepath_or_buffer, kwds):
 
     compression = kwds.get('compression', 'infer')
     compression = _infer_compression(filepath_or_buffer, compression)
-    filepath_or_buffer, _, compression, should_close = get_filepath_or_buffer(
+
+    # TODO: get_filepath_or_buffer could return
+    # Union[FilePathOrBuffer, s3fs.S3File, gcsfs.GCSFile]
+    # though mypy handling of conditional imports is difficult.
+    # See https://github.com/python/mypy/issues/1297
+    fp_or_buf, _, compression, should_close = get_filepath_or_buffer(
```
Comment (WillAyd) [inline pandas/io/parsers.py]: It's mentioned in the comments but I changed the variable name here from `filepath_or_buffer` to `fp_or_buf` to intentionally NOT shadow the parameter from the signature. 

As mentioned in the comment this local variable could potentially introduce new types for S3 and GCP and I don't think there is a great way with typing to statically analyze conditional imports like those just yet, so it's a clearer delimitation IMO to assign the return of this function to a separate variable

**Comment by jreback** [comment]
see some suggestions: https://github.com/pandas-dev/pandas/pull/25829 (can be a followup)

**Comment by jreback** [comment]
looks fine. better to start small. merge if you like (or add more up 2 you)

File: doc/source/whatsnew/v0.25.0.rst [hunk 3]
```diff
@@ -18,6 +18,18 @@ What's New in 0.25.0 (April XX, 2019)
...
 
+Enhancements
+~~~~~~~~~~~~
+
+.. _whatsnew_0250.enhancements.typing:
+
+Type Hints and ``pandas.typing``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In accordance with :pep:`484` pandas has introduced Type Hints and a new ``pandas.typing`` module containing aliases for idiomatic pandas types into the code base. We will be continually adding annotations to the code base to improve readability, reduce code maintenance and proactively identify bugs.
```
[the remark this answers is not in the document]
Comment (WillAyd) [reply doc/source/whatsnew/v0.25.0.rst]: Yea I don't see a downside to making private. Can do that on next iteration

[the remark this answers is not in the document]
Comment (WillAyd) [reply doc/source/whatsnew/v0.25.0.rst]: OK just made this private and updated documentation - lmk what you think

File: doc/source/whatsnew/v0.25.0.rst [hunk 4]
```diff
@@ -18,6 +18,18 @@ What's New in 0.25.0 (April XX, 2019)
 These are the changes in pandas 0.25.0. See :ref:`release` for a full changelog
 including other versions of pandas.
 
+Enhancements
+~~~~~~~~~~~~
+
+.. _whatsnew_0250.enhancements.typing:
+
+Type Hints and ``pandas.typing``
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: i would remove this entirely. if its private it is not to be relied upon.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_typing.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+from pathlib import Path
```
Comment (jreback) [inline pandas/_typing.py]: can you add a test that asserts this in pandas/tests/types/test_api.py

**Review by jreback — CHANGES_REQUESTED** [review]
needs a tests as indicated, other lgtm. ping on green.

File: pandas/_typing.py (hunk 5 above)
Comment (WillAyd) [reply pandas/_typing.py]: Yea no problem. So the existing test(s) exclude any privately named modules:

https://github.com/pandas-dev/pandas/blob/27927056b368fa8c9501c4ec8729546dbe98068c/pandas/tests/api/test_api.py#L14

Are you asking to revisit that logic or simply add a test with this as an exception to make sure it lives there?

Comment (jreback) [reply pandas/_typing.py]: oh i c. nvm then. though I think we should actually check and lock down all modules (so add back the private ones). in a new PR / issue though.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
merging this so can start using it. less concerned about the test as its private.

**Pull request merged**